events/listener.go

```diff
@@ -0,0 +1,675 @@
+// listener.go
+
+package events
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/chain4travel/camino-messenger-contracts/go/contracts/bookingtoken"
+	"github.com/chain4travel/camino-messenger-contracts/go/contracts/cmaccount"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+)
+
+// EventHandler defines the type for custom event handlers
+type EventHandler func(event interface{})
+
+// ListenerHandle is a handle returned when registering an event handler,
+// which can be used to stop listening to that event
+type ListenerHandle interface {
+	Unsubscribe()
+}
+
+// subscriptionInfo holds information about a subscription
+type subscriptionInfo struct {
+	subID       string
+	sub         event.Subscription
+	contract    common.Address
+	closeChan   func()
+	handler     EventHandler
+	unsubscribe context.CancelFunc // Used to cancel the resubscription
+}
+
+// EventListener listens for events from CMAccount and BookingToken contracts
+type EventListener struct {
+	client        *ethclient.Client
+	logger        *zap.SugaredLogger
+	mu            sync.Mutex
+	cmAccounts    map[common.Address]*cmaccount.Cmaccount
+	btContracts   map[common.Address]*bookingtoken.Bookingtoken
+	subscriptions map[string]*subscriptionInfo // Keyed by unique IDs
+}
+
+// NewEventListener creates a new EventListener instance
+func NewEventListener(client *ethclient.Client, logger *zap.SugaredLogger) *EventListener {
+	return &EventListener{
+		client:        client,
+		logger:        logger,
+		cmAccounts:    make(map[common.Address]*cmaccount.Cmaccount),
+		btContracts:   make(map[common.Address]*bookingtoken.Bookingtoken),
+		subscriptions: make(map[string]*subscriptionInfo),
+	}
+}
+
+// getOrCreateCMAccount gets or creates a CMAccount instance
+func (el *EventListener) getOrCreateCMAccount(addr common.Address) (*cmaccount.Cmaccount, error) {
```

Comment:
is it possible, that this func will be called concurrently? if yes, add rw lock

Reply:
Yes, updated.